V8 feature/304 support custom button

tjololo opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Introduce actions endpoint for use by custom buttons

## Related Issue(s)
- #304 

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

tjololo committed:
Make multi decision request for actions

tjololo committed:
Prepare code for useraction in task

tjololo committed:
add controller endpoint

tjololo committed:
trigger userdefined task actions

tjololo committed:
Add authorization check

tjololo committed:
Add some tests

tjololo committed:
controller tests for actions

tjololo committed:
added more tests for actionscontroller

tjololo committed:
Change how errors are returned

tjololo committed:
Merge branch 'v8' into v8-feature/304-support-custom-button

# Conflicts:
#	src/Altinn.App.Api/Controllers/ProcessController.cs

tjololo committed:
Add tests for MultiDecisionHelper and add xml comments

tjololo committed:
Add tests for AuthorizationClient

tjololo committed:
Fix some codesmells and add some more tests

tjololo committed:
Fix style error

tjololo committed:
Fix code smells

tjololo committed:
Fix code style

tjololo committed:
Remove some more code smells

tjololo committed:
fix yet another style issue

RonnyB71 commented on src/Altinn.App.Core/Features/Action/UserActionFactory.cs:
```diff
@@ -21,14 +23,30 @@ public UserActionFactory(IEnumerable<IUserAction> actionHandlers)
+            return _actionHandlers.FirstOrDefault(ah => ah.Id.Equals(actionId, StringComparison.OrdinalIgnoreCase));
+        }
+
+        return null;
```
Do we want to return null here or the NullUserAction?

RonnyB71 replied:
Noticed the next method where you can specify the defaultAction so I guess that should be used in cases where you would like to have the NullUserAction

tjololo replied:
Opted to return null here so that the aciton controller could return 404 if the user is authenticated to perform an action with a given name, but there is no registered service with that id. Think that would make debugging a typo easier.
The process engine defaults to the NullUserAction as there might not be a need for custom code just an action to make a decision in a following gateway

tjololo replied:
Noticed that the controller didn't return 404. Have added the code for that with a test

RonnyB71 commented on src/Altinn.App.Core/Helpers/MultiDecisionHelper.cs:
```diff
@@ -0,0 +1,186 @@
+/// <summary>
+/// Helper class for multi decision requests.
+/// </summary>
+public static class MultiDecisionHelper
```
Nice job 👏

RonnyB71 commented on src/Altinn.App.Core/Helpers/MultiDecisionHelper.cs:
```diff
@@ -0,0 +1,186 @@
+        string? task = instance.Process?.CurrentTask?.ElementId;
+        string instanceOwnerPartyId = instance.InstanceOwner.PartyId;
+        string org = instance.Org;
+        string app = instance.AppId.Split("/")[1];
```
Use AppIdentifier and avoid leaking the split logic and possibly return the AppIdentifier instead of Org and App?

tjololo replied:
Thanks! :+1:

RonnyB71 commented on src/Altinn.App.Core/Models/UserAction/FrontendAction.cs:
```diff
@@ -0,0 +1,33 @@
+    {
+        var frontendAction = new FrontendAction()
+        {
+            Name = "gotoPage",
```
Is this the one we need now, or is it navigateNextPage?

tjololo replied:
Correct we need the navigateToNextPage, will replace gotoPage with navigateToNextPage.

tjololo replied:
After having a quick look at the frontend issue the correct name is `nextPage`

RonnyB71 reviewed (approved):
🕹️

tjololo committed:
Action Controller return 404 if no implementations found
Fix comments from review

ivarne commented on src/Altinn.App.Core/Infrastructure/Clients/Authorization/AuthorizationClient.cs:
```diff
@@ -3,24 +3,23 @@
 namespace Altinn.App.Core.Infrastructure.Clients.Authorization
 {
-    /// <summary>
+        /// <summary>
```
```suggestion
    /// <summary>
```

ivarne commented on src/Altinn.App.Core/Infrastructure/Clients/Storage/ProcessClient.cs:
```diff
@@ -46,7 +46,7 @@ public class ProcessClient : IProcessClient
         public Stream GetProcessDefinition()
         {
-            string bpmnFilePath = _appSettings.AppBasePath + _appSettings.ConfigurationFolder + _appSettings.ProcessFolder + _appSettings.ProcessFileName;
+            string bpmnFilePath = Path.Combine(_appSettings.AppBasePath , _appSettings.ConfigurationFolder , _appSettings.ProcessFolder , _appSettings.ProcessFileName);
```
```suggestion
            string bpmnFilePath = Path.Join(_appSettings.AppBasePath , _appSettings.ConfigurationFolder , _appSettings.ProcessFolder , _appSettings.ProcessFileName);
```
https://learn.microsoft.com/en-us/dotnet/api/system.io.path.combine?view=net-8.0#remarks

> This method is intended to concatenate individual strings into a single string that represents a file path. However, if an argument other than the first contains a rooted path, any previous path components are ignored, and the returned string begins with that rooted path component. As an alternative to the Combine method, consider using the [Join](https://learn.microsoft.com/en-us/dotnet/api/system.io.path.join?view=net-8.0) or [TryJoin](https://learn.microsoft.com/en-us/dotnet/api/system.io.path.tryjoin?view=net-8.0) methods.

ivarne commented on src/Altinn.App.Api/Controllers/ActionsController.cs:
```diff
@@ -0,0 +1,130 @@
+                UpdatedDataModels = result.UpdatedDataModels
+            });
+        }
+        catch (NotFoundException)
```
Do we need a custom exception here? Can't `_userActionService` just return `null` if action is not found, or just add some magic in `UserActionResponse` so that it can indicate the HTTP status code (either through a `statusCode` property in `ActionError`, or by looking at the `Code` property and change the status if it is `ActionNotFound`)

tjololo replied:
Actually the whole `UserActionService` could be removed now. The main purpose of that service was to add the, now removed, validationIssues to a validation group. As this concept is removed we can remove the service and handle the whole action not found in the controller without using exceptions, and remove a service.
I will rewrite the code

ivarne replied:
If you rewrite, I'd rather you merge `UserActionFactory` into `UserActionService` instead of merging the service into the controller.

ivarne commented on src/Altinn.App.Api/Models/UserActionRequest.cs:
```diff
@@ -0,0 +1,27 @@
+    /// <summary>
+    /// The id of the button that was clicked
+    /// </summary>
+    public string? ButtonId { get; set; }
```
```suggestion
    [JsonPropertyName("buttonId")]
    public string? ButtonId { get; set; }
```

ivarne commented on test/Altinn.App.Api.Tests/Utils/PrincipalUtil.cs:
```diff
@@ -7,21 +7,25 @@ namespace Altinn.App.Api.Tests.Utils
+                claims.Add(new Claim(AltinnCoreClaimTypes.UserId, userId.ToString()!, ClaimValueTypes.String, issuer));
+                claims.Add(new Claim(AltinnCoreClaimTypes.PartyID, userId.ToString()!, ClaimValueTypes.Integer32, issuer));
+            }
+
```
What problem does this solve?

tjololo replied:
`ClaimsPrincipalExtensions.GetUserIdAsInt` can return null for userid if the `urn:altinn:userid` is not set in the token. To test that path in ActionsController I marked userId as nullable and a if statement to not add the userClaims if userId is null

ivarne replied:
Will have to look at this when https://github.com/Altinn/app-lib-dotnet/pull/353 and this is done.

ivarne reviewed (commented):
Tror muligens det er en del filer som mangler `[JsonPropertyName("")]`

tjololo committed:
Apply first round of suggestions from code review

Co-authored-by: Ivar Nesje <ivarne@users.noreply.github.com>

tjololo committed:
Apply suggestion from code review

Remove IUserActionService and rename UserActionFactory to UserActionService
Add JsonPropertyName to classes returned through the ActionsController

tjololo committed:
Remove ValidationGroup from IUserAction

ivarne commented on src/Altinn.App.Core/Internal/Process/ProcessEngine.cs:
```diff
@@ -131,7 +131,7 @@ public async Task<ProcessChangeResult> Next(ProcessNextRequest request)
             };
         }
-        var actionHandler = await _userActionFactory.GetActionHandlerOrDefault(request.Action, new NullUserAction()).HandleAction(new UserActionContext(request.Instance, userId.Value));
+        var actionHandler = await _userActionService.GetActionHandlerOrDefault(request.Action, new NullUserAction()).HandleAction(new UserActionContext(request.Instance, userId.Value));
```
Is it actually correct to have an action with no implementation, so that the `NullUserAction` is required?

tjololo replied:
You can use the action Id to navigate the process without having any custom logic connected to it. In those cases I think we shouldn't force the developer to implement a empty IUserAction and just default to the `NullUserAction`

ivarne replied:
In that case I'd rather check the null return value from `UserActionServcie.GetActionHandler`, so the code explicitly states how to handle a case without custom logic.

ivarne commented on test/Altinn.App.Core.Tests/Features/Action/UserActionServiceTests.cs:
```diff
@@ -0,0 +1,97 @@
+    
+    internal class DummyUserAction : IUserAction
+    {
+        public string Id { get; set; } = "dummy";
```
This means the same, but is prettier
```suggestion
        public string Id => "dummy";
```

ivarne commented on test/Altinn.App.Core.Tests/Features/Action/UserActionServiceTests.cs:
```diff
@@ -0,0 +1,97 @@
+    
+    internal class DummyUserAction2 : IUserAction
+    {
+        public string Id { get; set; } = "dummy";
```
```suggestion
        public string Id => "dummy";
```

ivarne commented on src/Altinn.App.Core/Internal/Process/Elements/UserAction.cs:
```diff
@@ -0,0 +1,31 @@
+        /// Gets or sets the ID of the action
+        /// </summary>
+        [JsonPropertyName("id")]
+        public string Id { get; set; }
```
I'm itching in my fingers to add this everywhere. Just curious when it will cause bugs.
```suggestion
        public required string Id { get; set; }
```

ivarne commented on src/Altinn.App.Core/Models/UserAction/ActionError.cs:
```diff
@@ -0,0 +1,27 @@
+    /// Machine readable error code
+    /// </summary>
+    [JsonPropertyName("code")]
+    public string Code { get; set; } = string.Empty;
```
Could this also be `required` instead of setting a default?

tjololo replied:
I'm not sure, our frontend code would probably not read this field, but it is meant for those applications with api consumers. Should we force apps without API consumers to set a field that is never used?

ivarne replied:
That is why I'm asking. If it is optional, it might make more sense to have it nullable?

tjololo replied:
good point, lets set it as nullable 👍

ivarne reviewed (approved):
Just fix the code smell/warning, and I think this code is ready to run and find bugs runtime instead.

tjololo committed:
Apply suggestions from code review

Co-authored-by: Ivar Nesje <ivarne@users.noreply.github.com>

tjololo committed:
Remove the need for NullUserAction

ivarne commented:
🕹️
